### PR TITLE
SP-1935 GearMenu:

### PR DESCRIPTION
**Pull request by hsuchie4**
- Gear Menu Security table added into Security lists
- Gear Menu Security button requires new icon/image (currently using link security)
- Gear Menu Security table functional, admin can select which menus of which user group can see under the gear menu

### CODE REVIEW

File: src/tactic/ui/startup/security_wdg.py [hunk 1]
```diff
@@ -1608,7 +1632,79 @@ def get_sobjects(my, group_names):
...
+                               'Tasks': {'label': ['Show Tasks','Add Tasks to Selected','Add Tasks to Matched'], 'order': 7},
+                               'Notes': {'label': ['Show Notes'], 'order': 8},
+                               'Check-ins': {'label': ['Show Check-in History'], 'order': 9},
+                               'Pipelines': {'label': ['Show Pipeline Code','Edit Pipelines'], 'order': 10}
+                               }
+        all_gear_menu_names = sorted(all_gear_menu_names.iteritems(), key=lambda (x,y):y['order'])
+        return all_gear_menu_names
+
+    
+    get_menu = classmethod(get_all_menu_names)
```
Comment (oboreo) [inline src/tactic/ui/startup/security_wdg.py]: this should be the same name "get_all_menu_names"

File: src/tactic/ui/startup/security_wdg.py [hunk 2]
```diff
@@ -1608,7 +1632,79 @@ def get_sobjects(my, group_names):
...
+    
+
+    def get_sobjects(my, group_names):
+
+        all_gear_menu_names = GearMenuSecurityWdg.get_all_menu_names(my)
```
Comment (oboreo) [inline src/tactic/ui/startup/security_wdg.py]: replace (my)  with()

File: src/tactic/ui/widget/gear_menu_wdg.py [hunk 3]
```diff
@@ -79,8 +79,30 @@ def get_args_keys(my):
             "embedded_table": "boolean to show if it is part of an embedded table"
         }
 
+    def get_access_keys_dict(my):
 
-    def add_custom_menus( my, menus, menu_idx_map, custom_menus ):
+        from tactic.ui.startup import GearMenuSecurityWdg
+        get_menu = GearMenuSecurityWdg.get_menu()
```
Comment (oboreo) [inline src/tactic/ui/widget/gear_menu_wdg.py]: it should be called whatever name you have decided in the other file (not get_menu())

File: src/tactic/ui/startup/security_wdg.py [hunk 4]
```diff
@@ -1608,7 +1632,79 @@ def get_sobjects(my, group_names):
...
+        rules_dict = {}
+
+        project = Project.get()
+        project_code = project.get_code()
+        
+        sobjects = []
+
+
+        for key in all_gear_menu_names:
```
Comment (oboreo) [inline src/tactic/ui/startup/security_wdg.py]: for k,v in d.items():
    print k, 'corresponds to', v

File: src/tactic/ui/startup/security_wdg.py [hunk 5]
```diff
@@ -2228,13 +2363,58 @@ def remove_project(my, project_code):
...
+        if (submenu, label) not in (submenus, labels):
+            rule = my.xml.create_element("rule")
+            my.xml.set_attribute(rule, "group", "gear_menu")
+            my.xml.set_attribute(rule, "submenu", submenu)
+            my.xml.set_attribute(rule, "label", label)
+            if project_code:
+                my.xml.set_attribute(rule, "project", project_code)
+            my.xml.set_attribute(rule, "access", access)
+            my.xml.append_child(my.root, rule)
+            '''
```
Comment (oboreo) [inline src/tactic/ui/startup/security_wdg.py]: remove commented code

File: src/tactic/ui/startup/security_wdg.py [hunk 6]
```diff
@@ -1608,7 +1632,79 @@ def get_sobjects(my, group_names):
...
+                               'Clipboard': {'label': ['Copy Selected','Paste','Connect','Append Selected','Show Clipboard Contents'], 'order': 3},
+                               'View': {'label': ['Column Manager','Create New Column','Save Current View','Save a New View','Edit Current View','Edit Config XML'], 'order': 4},
+                               'Print': {'label': ['Print Selected','Print Displayed','Print Matched'], 'order': 5},
+                               'Chart': {'label': ['Chart Items','Chart Selected'], 'order': 6},
+                               'Tasks': {'label': ['Show Tasks','Add Tasks to Selected','Add Tasks to Matched'], 'order': 7},
+                               'Notes': {'label': ['Show Notes'], 'order': 8},
+                               'Check-ins': {'label': ['Show Check-in History'], 'order': 9},
+                               'Pipelines': {'label': ['Show Pipeline Code','Edit Pipelines'], 'order': 10}
+                               }
+        all_gear_menu_names = sorted(all_gear_menu_names.iteritems(), key=lambda (x,y):y['order'])
```
Comment (oboreo) [inline src/tactic/ui/startup/security_wdg.py]: .items()

File: src/tactic/ui/widget/gear_menu_wdg.py [hunk 7]
```diff
@@ -344,7 +486,7 @@ def get_file_menu(my):
                     '''%my.ingest_data_view
                  }
                 } )
-
+        if security.check_access("builtin", access_keys, "allow") or 'Check-out Files' in label_list:
```
Comment (oboreo) [inline src/tactic/ui/widget/gear_menu_wdg.py]: remove 
security.check_access("builtin", access_keys, "allow") or

File: src/tactic/ui/startup/security_wdg.py [hunk 8]
```diff
@@ -1608,7 +1632,79 @@ def get_sobjects(my, group_names):
 
         return sobjects
 
+class GearMenuSecurityWdg(ProjectSecurityWdg):
+
```
Comment (oboreo) [inline src/tactic/ui/startup/security_wdg.py]: '''Control Gear Menu Visibility'''

File: src/tactic/ui/startup/security_wdg.py [hunk 9]
```diff
@@ -1608,7 +1632,79 @@ def get_sobjects(my, group_names):
...
+    def get_security_type(my):
+        return "gear_menu"
+
+    def get_save_cbk(my):
+        return 'tactic.ui.startup.GearMenuSecurityCbk'
+
+    def get_display_columns(my):
+        return ['submenu', 'label']
+
+    def get_all_menu_names(my):
```
Comment (oboreo) [inline src/tactic/ui/startup/security_wdg.py]: replace my with cls

File: src/tactic/ui/widget/gear_menu_wdg.py (hunk 7 above)
Comment (hsuchie4) [inline src/tactic/ui/widget/gear_menu_wdg.py]: @oboreo so did you want me to remove this still or not?

**Pull request merged**